File: ocaml/2015/day06a/day06a.ml
open Printf
open Sys

let usage progname =
  eprintf "usage: %s <input file>\n" progname;
  exit 1

let perform grid action (r1, c1, r2, c2) =
  for row = r1 to r2 do
    for col = c1 to c2 do
      match action with
      | "turn on" -> grid.(row).(col) <- true
      | "turn off" -> grid.(row).(col) <- false
      | "toggle" -> grid.(row).(col) <- (not grid.(row).(col))
      | _ -> ()
    done
  done

let count grid =
  let count = ref 0 in
  for row = 0 to 999 do
    for col = 0 to 999 do
      if grid.(row).(col) then
        count := !count + 1
      else
        ()
     done
  done ;
  !count

let process_file filename =
  try
    let grid = Array.init 1000 (fun _ -> Array.make 1000 false) in
    let lines = In_channel.input_lines (In_channel.open_text filename) in
    let re = Str.regexp {|\(turn on\|turn off\|toggle\) \([0-9]+\),\([0-9]+\) through \([0-9]+\),\([0-9]+\)|} in
    Seq.iter (fun line ->
      if Str.string_match re line 0 then
        let action = Str.matched_group 1 line in
        let r1 = int_of_string (Str.matched_group 2 line) in
        let c1 = int_of_string (Str.matched_group 3 line) in
        let r2 = int_of_string (Str.matched_group 4 line) in
        let c2 = int_of_string (Str.matched_group 5 line) in
        perform grid action (r1, c1, r2, c2)
    ) (List.to_seq lines) ;
    count grid
  with
  | Sys_error msg ->
      eprintf "error reading file: %s\n" msg;
      exit 1

let () =
  let args = Sys.argv
  in if Array.length args < 2 then
       usage args.(0) ;
     let filename = args.(1) in
     let result = process_file filename in
     printf "result = %d\n" result

File: nim/2015/day06a/day06a.nim
import os, re, strutils

proc perform(grid: var array[1000*1000, bool], action: string, r1: int, c1: int, r2: int, c2: int) =
  for row in r1..r2:
    for col in c1..c2:
      if action == "turn on":
        grid[row*1000 + col] = true
      elif action == "turn off":
        grid[row*1000 + col] = false
      elif action == "toggle":
        grid[row*1000 + col] = not grid[row*1000 + col]

proc count(grid: array[1000*1000, bool]): int =
  var total = 0
  for row in 0..999:
    for col in 0..999:
      if grid[row*1000 + col]:
        total += 1
  total

proc processFile(filename: string): int =
  var grid: array[1000*1000, bool]
  var caps: array[5, string];
  for line in lines(filename):
    if find(line, re"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", caps) != -1:
      let action = caps[0]
      let r1 = parseInt(caps[1])
      let c1 = parseInt(caps[2])
      let r2 = parseInt(caps[3])
      let c2 = parseInt(caps[4])
      perform(grid, action, r1, c1, r2, c2)
  count(grid)

when isMainModule:
  if paramCount() < 1:
    quit "usage: " & getAppFilename() & " <input file>", 1

  let filename = paramStr(1)
  try:
    echo "result = ", processFile(filename)
  except OSError as e:
    quit "error while processing file `" & filename & "': " & e.msg, 1
